static/js/stream_edit.js

```diff
@@ -63,6 +63,51 @@ exports.get_users_from_subscribers = function (subscribers) {
     });
 };
 
+exports.get_retention_policy_text_for_subscription_type = function (sub) {
+    let message_retention_days = sub.message_retention_days;
+    if (page_params.realm_message_retention_days === -1 && (sub.message_retention_days === null ||
+            sub.message_retention_days === -1)) {
+        return;
+    }
+
+    if (sub.message_retention_days === -1) {
+        return i18n.t("Messages in this stream will be retained forever.");
+    }
+
+    if (message_retention_days === null)  {
+        message_retention_days = page_params.realm_message_retention_days;
+    }
+
+    return i18n.t("Messages in this stream will be automatically deleted after __retention_days__ days.", {retention_days: message_retention_days});
+};
+
+exports.get_display_text_for_realm_message_retention_setting = function () {
+    if (page_params.realm_message_retention_days === -1) {
+        return i18n.t("(forever)");
+    }
+    return i18n.t("(__message_retention_days__ days)", {message_retention_days: page_params.realm_message_retention_days});
```

Comment:
`__message_retention_days__` will be rendered as empty when `page_params.realm_message_retention_days` is `null`.

Reply:
In my PR, https://github.com/zulip/zulip/pull/14938, we have a pending discussion of what should be the best approach to set the `message_retention_days` to "Forever".

In master, both `null`, as well as `-1`, means "Forever", but this is not coherent with Streams case where `-1` means "disable retention policy for this stream unconditionally", so setting `-1` at the realm level implies "No matter what we want the message to retain forever even if some streams have different value".

Reply:
Message retention docs [here](https://zulipchat.com/help/message-retention-policy) says 
`Setting a retention policy for individual streams, to just delete messages on those specific streams.` and it suggests that we will delete the messages on the streams even if realm setting is set to forever.

Reply:
... which is bit confusing.

I see how you managed the "Forever" case for the stream by passing a string "forever" rather than sticking to a number, @timabbott @mateuszmandera what are your opinions on this?

This makes me wonder, at the realm level rather than "Retain forever" it should mean "Retain forever unless specified at stream level"?

Reply:
I also have gone through some code in `zerver/lib/retention.py` and we do not handle case with realm value set to -1. It seems like in this file we assume that null means forever.

Reply:
> Message retention docs [here](https://zulipchat.com/help/message-retention-policy) says
> `Setting a retention policy for individual streams, to just delete messages on those specific streams.` and it suggests that we will delete the messages on the streams even if realm setting is set to forever.

(GitHub real-time page update is kinda broken, that's why I didn't notice you reply)

yes, that's true, I was saying from a developers perspective and maybe from API standpoint too, that `-1` at `Stream` side means something else. Related discussion: https://github.com/zulip/zulip/pull/14903#issuecomment-626204683

Reply:
Yeah -1 for stream means we will retain forever inspite of realm-setting value being a finite postive number.
I was thinking the same as you mentioned in that comment, that we could use None as retain forever rather than using -1. Maybe we can wait on decision of whether to use -1 or None for the forever case in realm-level settings

Reply:
The confusion is due to the fact that `Realm.message_retention_days` does not mean a strict "keep messages forever".

`Realm.message_retention_days` being NULL means the natural thing "no realm-wide policy set", which is the default situation and we're not archiving old messages. So then, if a stream sets it retention policy then we're going to archive the stream's message according to that - there's no contradiction with the realm's policy when the realm has no policy.

Streams have the special option of declaring "keep my messages forever" (through the value `-1`), because that's something potentially useful if the admin wants to exclude certain streams from the realm-wide policy.  The equivalent doesn't make sense for an entire realm, a realm can merely have no policy declared (`NULL`)

Does that help clear things up? :thinking:

Reply:
Thanks, it does makes a lot of things clear to me. 
So, we are changing realm-wide policy to be NULL for forever case and are removing use of -1 as implemented in #14938, correct?

Reply:
So far we're not changing the meaning of any values for the `message_retention_days` field on the backend models. The problem in the PR is "how to convey `no policy set` to the server through the API" - because the webapp can't send the `null` value, because that already has a meaning in the setting system and means "no change to this setting has been made". So for now in the PR's code `-1` is used to represent that. But that gets very confusing once streams come into play as you can see...

Reply:
Okay I just got caught up on https://chat.zulip.org/#narrow/stream/2-general/topic/message.20retention.20for.20streams/near/903209 So if @sahil839 is going with string values, so should #14938, instead of sending the `-1`, this is much better.